### ISSUE TITLE
Mail translation change the product table location in html template.

### ISSUE THREAD

**Issue by Yuraken** [PrestaShop/PrestaShop#21630]
#### Describe the bug

Whenever I translate our mail from English to French, our product table is moved in the template. 

#### Expected behavior

I expect the mail translator from prestashop to translate the translable variables on my template without messing up the template.

#### Steps to Reproduce

1. Generate mordern mail templates
2. Go to translation page for mail with modern template.
3. Go on the order_conf mail template and translate a word.
4. Save.
5. Go to mail theme page. 
6. Select modern template.
7. Select order_conf template and send a test mail.
8. You can see that the product table is not at the same place as it was in the twig template before translation.

**Screenshots**
Mail before translation : 
<img width="443" alt="2020-10-27_09h36_31" src="https://user-images.githubusercontent.com/32276415/97277094-59dfb380-1838-11eb-9e67-c221929b8cf1.png">

Mail after translation : 
<img width="457" alt="2020-10-27_09h38_20" src="https://user-images.githubusercontent.com/32276415/97277151-6a902980-1838-11eb-9bea-19a8bc4c882e.png">

#### Additional information

* PrestaShop version: 1.7.6.3
* PHP version: 7.2

## SPECS

We can remove the WYSIWYG editor for HTML template in International > Translations as the content of the email can still be edited through Improve > Design > Email Theme > Translate emails, while waiting to give again the possibility to customize email's themes properly

**Comment by JS-starlord** [comment]
Same here, everytime I try to translate order_conf , the html generated at the end will be broken. It seem that the {product} variable broke the template : it move the product list outside the `<TABLE>`  in the HTML code ...

**Comment by khouloudbelguith** [comment]
Hi @Yuraken,

I manage to reproduce the issue with PS1.7.6.8 and PS1.7.7.0rc1build2.
![image](https://user-images.githubusercontent.com/16067358/97285086-11c58e80-1842-11eb-8d46-932717ed1cc9.png)
This is a new feature 1760, it was ok with email theme = `classic` (PS1.7.5.2).
I’ll add this to the debug backlog so that it’s fixed.
If you have already fixed it on your end or if you think you can do it, please do send us a pull request!

Thank you

**Comment by atomiix** [comment]
@PrestaShop/product-team I tried something to fix this issue but as discussed with @eternoendless it's far from optimal and it should not be solved that way. So, to summarize what's happening:

The way we can edit email translations by modifying the email template through *Improve > International > Translations* is only partially working. Because we're in a WYSIWYG editor and email templates can't be rendered entirely, when saving, we may break the HTML code, leading to broken email templates once sent.

#11882 is what we planed to do but in the meanwhile, we may have to find another solution (like disabling the WYSIWYG editor) unless we decide that it should be developed for 1.7.7.x. WDYT?

**Comment by marionf** [comment]
@atomiix If we disable the WYSIWYG editor, it will no longer be possible to edit the HTML version of the emails through International > Translations ?

**Comment by atomiix** [comment]
@marionf exactly. You can change translations through *Improve* > *Design* > *Email Theme* > *Translate emails* though, but not the html structure.

**Comment by marionf** [comment]
As this is not a regression of the 1.7.7.0 (this issue is existing since 1.7.6.0 where we introduced the new email system with modern templates), I remove it from the 1.7.7.x kanban and add it to the 1.7.8 one.

We can remove the WYSIWYG editor for HTML template in _International > Translations_ as the content of the email can still be edited through _Improve > Design > Email Theme > Translate emails_,  while waiting to give again the possibility to customize email's themes properly

**Comment by promatik** [comment]
I had this with `v1.7.6.8`, I thought by updating to the latest `v1.7.7.7` the issue would be solved, but it isn't.

Is there any way to fix this locally?

**Comment by eternoendless** [comment]
I think you might fix this if you translate the email wordings from _Design > Email theme > Translate_

**Comment by ballashop** [comment]
Hi,
the problem comes back to us like this:
if we choose the default English language the **order_config** email template is sent with the correct layout
if we switch from English to Italian language, the layout has broken as indicated above.
So only by switch the language, without changes anything inside the e-mail code, the order_conf change layout.
Thanks any reply.
Balla

I add... the emails that arrive to the site administrator have the correct layout, in both languages. So this problem only affects the email sent to the customers.

**Comment by khlayel313** [comment]
Hi everoyone,

Same bug here in prestashop 1.7.8.7

Any solutions please ?

**Comment by khlayel313** [comment]
As solution that i find is to edit email templates directly from files in /themes/my_theme/mails/fr/order_conf.html.

Tested and works perfectly

**Comment by ZIT2050** [comment]
Oh My God, today is 2023-04-05 and I am "discovering" this issue on PrestShop 1.7.8.8 released on **December 2022** and see this bug is open from **October 2020**. So over 2 years and nobody fix this?

I spent many hours thinkig what I am doing wrong ...

**Comment by eternoendless** [comment]
@ZIT2050 have you tried this? https://github.com/PrestaShop/PrestaShop/issues/21630#issuecomment-945765835

**Comment by ZIT2050** [comment]
Thanks @eternoendless 
Of course I can manage with this - I think best solution is to _"change by hand" _ especially when you have child_theme in use. Also putting original *.html files can help

**Comment by boo-code** [comment]
This looks fixed on 9.x, and the mechanism is worth writing down because it explains why @atomiix said the earlier pull request would not solve it.

The mail body is edited through TinyMCE, and the moved table is not TinyMCE filtering elements — its configuration is already as permissive as it gets, `valid_elements: '*[*]'` and `valid_children: '+*[*]'`. It is the **HTML parser**. A bare `{placeholder}` sitting directly inside table markup is a text node in a place HTML does not allow text, so the parser foster-parents it out of the table, and the surrounding structure moves with it. That happens before any TinyMCE schema applies, which is why fixing the render does not help.

`translation_mails.tpl` on 9.2.x now guards exactly that:

```js
rte_mail_config['protect'] = [/(?<=<\/(?:tr|thead|tbody|tfoot)>\s*(?:\{[a-zA-Z_]\w*\}\s*)*)\{[a-zA-Z_]\w*\}/g];
```

TinyMCE's `protect` swaps those matches for comment placeholders before parsing, so they survive the round trip.

Checked against the shipped template rather than assumed. In `mails/en/order_conf.html` there are 24 placeholders, and exactly two sit in a position the parser would foster-parent:

| placeholder | position | covered by the guard |
| --- | --- | --- |
| `{products}` | after `</tr>` | yes |
| `{discounts}` | after `</tr>` | yes |

Those two are the product table and the discount table, which is what this issue reports.

That line is on `9.2.x` and **not** on `8.2.x`, where the config block ends at `plugins`. So the fix is 9.x only.

One limit worth stating, for anyone with a custom template: the guard covers placeholders that follow a `</tr>`, `</thead>`, `</tbody>` or `</tfoot>`. A placeholder in another foster-parenting position, for instance directly after `<table>` or `<tbody>` and before the first `<tr>`, is still moved. The shipped templates do not do that, but a custom one can.

Suggest QA re-verifies on a current 9.x version and closes.
